test/src/test/java/org/corfudb/infrastructure/LayoutBuilderTest.java

```diff
@@ -418,6 +402,61 @@ public void unsuccessfulSequencerFailoverThrowsException() throws Exception {
                 .addUnresponsiveServers(Collections.singleton(SERVERS.ENDPOINT_1))
                 .addUnresponsiveServers(Collections.singleton(SERVERS.ENDPOINT_2))
                 .assignResponsiveSequencerAsPrimary(Collections.singleton(SERVERS.ENDPOINT_0));
+
         layoutBuilder.build();
     }
+
+    /**
+     * This test verifies that when primary sequencer fails over, we try to assign
+     * a non log tail node as new primary sequencer if possible to achieve better
+     * load distribution.
+     */
+    @Test
+    public void assignPrimarySequencerToNonTailNodeIfPossible() {
+        Layout originalLayout = new TestLayoutBuilder()
+                .setEpoch(1L)
+                .addLayoutServer(SERVERS.PORT_0)
+                .addLayoutServer(SERVERS.PORT_1)
+                .addLayoutServer(SERVERS.PORT_2)
+                .addSequencer(SERVERS.PORT_0)
+                .addSequencer(SERVERS.PORT_1)
+                .addSequencer(SERVERS.PORT_2)
+                .buildSegment()
+                .buildStripe()
+                .addLogUnit(SERVERS.PORT_1)
+                .addLogUnit(SERVERS.PORT_2)
+                .addToSegment()
+                .buildStripe()
+                .addLogUnit(SERVERS.PORT_2)
+                .addLogUnit(SERVERS.PORT_0)
+                .addToSegment()
+                .buildStripe()
+                .addLogUnit(SERVERS.PORT_2)
+                .addToSegment()
+                .addToLayout()
+                .build();
+
+        // All servers responsive, server 1 is the only non-tail node.
+        LayoutBuilder layoutBuilder = new LayoutBuilder(originalLayout);
+        Layout newLayout = layoutBuilder
+                .assignResponsiveSequencerAsPrimary(Collections.emptySet())
+                .build();
+        assertThat(newLayout.getPrimarySequencer()).isEqualTo(SERVERS.ENDPOINT_1);
```

Comment:
I have run your test with the previous version and the current one, it seems that the previous version works even better.
It's not clear to me what improvement has been done if your test works the same way for both versions.

Regarding the check above, as I understand it must be ENDPOINT_0, so why you expect to see ENDPOINT_1 here?

Reply:
This is just a unit test not perf test. The purpose of the test is to make sure a non-tail node will be elected for primary sequencer if there is any. It also tests if all nodes are tails or all nodes are unresponsive, the behavior is same as before.

Your understanding is not correct, endpoint 0 is a tail node, so it won't be elected as primary sequencer in the algorithm. Endpoint 1 is a non-tail node, so it will be elected. The test might also work for old implementation because old implementation does not have any preference on which node being elected as long as they are responsive. So endpoint 0,1,2 are all possible. All my codes are documented, please read again.

Reply:
Ok, let's start from the beginning, questions, please answer:
 - The layout that you created in the test is not possible in production. We always have only one stripe in one segment. What was the motivation for that?
 
 - making Sequencer election depending on stripes is not the best idea, in my opinion, it makes logic pretty convoluted and unclear, why we need to overcomplicate that logic? 
    The Sequencer is a much more important component of the cluster than stripe and must not be dependent on a stripe. Maybe we can change the FIRST stripe server in the last segment depending on the primary sequencer but not vise versa.

Reply:
> The layout that you created in the test is not possible in production. We always have only one stripe in one segment. What was the motivation for creating multiple stripes in the test?

Striping is a design concept in corfu. We don't have multi-stipe in production does not mean we don't need to test it. You can see a lots of tests have multi-stripe layout. On the other hand, mutli-stripe test is a "super set" of single sripe, it's always better to make sure it works for multi-stripe, which implies it works for single stripe.

> making Sequencer election depending on stripes is not the best idea, in my opinion, it makes logic pretty convoluted and unclear, why we need to overcomplicate that logic?
The Sequencer is a much more important component of the cluster than stripe and must not be dependent on a stripe. Maybe we can change the FIRST stripe server in the last segment depending on the primary sequencer but not vise versa.

This is not depending on stripes, it just give high priority to non tail nodes, you can think of it as an optimization which takes advantage of the log tail information we have in the layout builder, it's not a dependency. Also I don't think the logic is convoluted, it's pretty clear: give high priority to non-tail nodes. You have a bunch of nodes to choose from, just give high priority to those who are not tails.
When you need to do re-election on sequencer, why do you want to change log unit?  You cannot arbitrarily change log units order, there needs to be more layout changes and state transfer.

Reply:
>it's not a dependency
It is a dependency, the design is - the more important component (sequencer) depends on a less important one (which is an optimization that not even fully used in production). 
I believe that kind of dependency is not acceptable.

Reply:
Please provide reason that makes it not acceptable. If it's because of "convoluted" than I already said it's actually pretty  simple, how could ten lines of change being convoluted. It's just a simple optimization, which will be downgraded to previous approach if all responsive nodes are tails.

> (which is an optimization that not even fully used in production).
This is used in production, which has a single node as tail, the optimization still applies to this case. The unit test verifies on a more general case.

Reply:
I believe that kind of dependency is not acceptable. 
The most important component of the cluster depends on the least important

Reply:
First this is not denpendency. We  are currently deploying everything on a node, but you can even choose to deploy sequencer on a node that does not have log unit server. this is the initial design. This patch does a similar thing.
By the way sequencer already depend on log unit server for fail-over (get persisted address map and stream tails.) Without this dependency, system won't work. Also log unit is not the least important, by design sequencer is just an optimization for log unit. Log unit is the source of truth.

Reply:
>We are currently deploying everything on a node, but you can even choose to deploy sequencer on a node that does not have log unit server. this is the initial design.

It has nothing to do with the deployment. We are talking only about layout. To figure out what server will be the next sequencer we need to have information about stripes.

>By the way sequencer already depend on log unit server for fail-over (get persisted address map and stream tails.)

could you please point where it is

>by design sequencer is just an optimization for log unit

Never heard about that, can you please provide a link to the initial design where it claimed?

>Log unit is the source of truth.

It's not true if we are talking about a particular node, that's why we have AutoCommitService and read the latest global address from the entire cluster, right?

Reply:
The deployment is an example. When deploying which node to be sequencer, system admin can decide to deploy a sequencer on non log unit node, which uses stripe information. By the way, you have stripe info in your layout why not use it to do optimization. Sequencer itself is an optmization to separate token issue responsibility from log unit. Log unit is the source of truth for tokens, it can issue tokens instead, we assign this responsibility to sequencer to get better performance. Ideally sequencer should be deployed on a sparate node to achieve most performance. When sequencer fails over it needs to query log unit for the persisted tokens, that means sequencer is an optimization for log unit, that dependency is there from day 1. You can take a look at LayoutManagementView::reconfigureSequencerServers.

Reply:
>By the way, you have stripe info in your layout why not use it to do optimization.

because I don't see it as optimization and how it can improve things.

>Sequencer itself is an optmization.

this is just your personal opinion

>When sequencer fails over it needs to query log unit for the persisted tokens
>that means sequencer is an optimization for log unit

Not clear to me how you get to the conclusion that sequencer is an optimization. 
And not clear how we can get from the above statements to the conclusion that the sequencer should be chosen from stripes.

>that dependency is there from day 1

Sequencer needs to know the latest global address, but the election process has nothing to do with the address

>You can take a look at LayoutManagementView::reconfigureSequencerServers.

could you please show the exact point in the code?

Reply:
> this is just your personal opinion

If you ever attended a new team member on board session, you will also hear maithem telling them sequencer is an optimization. Log unit is the persistent storage, that is the ground truth and most important thing.

> And not clear how we can get from the above statements to the conclusion that the sequencer should be chosen from stripes.

This is not to get that conclusion, it's to correct your statement that log unit is the least important.
Also it's not choosing from stripes, it's avoiding from choose a tail node, even there is no log unit server, no stripe, the algorithm still works. That's not a dependecy.

> Sequencer needs to know the latest global address, but the election process has nothing to do with the address

Failover includes election and recorver sequencer state from log unit, which already uses stripe info.

> could you please show the exact point in the code?

https://github.com/CorfuDB/CorfuDB/blob/master/runtime/src/main/java/org/corfudb/runtime/view/LayoutManagementView.java#L425-L426

Reply:
>it's to correct your statement that log unit is the least important.

One more time, stripe != log unit

>https://github.com/CorfuDB/CorfuDB/blob/master/runtime/src/main/java/org/corfudb/runtime/view/LayoutManagementView.java#L425-L426

the change was made 2 months ago
![image](https://user-images.githubusercontent.com/314292/95164533-01e4ee00-075f-11eb-85fb-0a15ba463c3d.png)
looks like also a false assumption

P.S. anyways, let's not spend so much time on an endless discussion. 
It's up to up to address or not my concerns

Reply:
> One more time, stripe != log unit

Every read/write request uses stripe info, I don't know how you come with the statement that stripe is not important. Maybe you think we have only 1 stripe deployment, but to make code work for general case, you need to consider stripping in your implentation.

> looks like also a false assumption

One more time, please read code carefully before commenting. getLogAddressSpace() calls getChainHeadFromAllSegments(), which uses stripe information.

> P.S. anyways, let's not spend so much time on an endless discussion.

I agree, this is kind of wasting time, if you need help in understanding how some certain components work in system, I can help offline.